File: pkg/store/store.go
// Package store is the physical storage service (bfs store module).
// It owns volumes and exposes add / get / del operations.
package store

import (
	"errors"
	"fmt"
	"path/filepath"
	"sync"

	"github.com/levis9527/jfs/pkg/meta"
	"github.com/levis9527/jfs/pkg/volume"
)

var (
	ErrVolumeNotFound = errors.New("store: volume not found")
	ErrNoWritable     = errors.New("store: no writable volume")
)

// Store manages one or more volumes on local disk.
type Store struct {
	mu      sync.RWMutex
	root    string
	vols    map[int32]*volume.Volume
	maxSize int64
}

// Config for store initialization.
type Config struct {
	DataDir     string
	VolumeIDs   []int32
	VolumeSize  int64
	VolumeCount int // if VolumeIDs empty, create 1..VolumeCount
}

// Open initializes the store and opens volumes.
func Open(cfg Config) (*Store, error) {
	if cfg.DataDir == "" {
		return nil, errors.New("store: DataDir required")
	}
	if cfg.VolumeCount <= 0 && len(cfg.VolumeIDs) == 0 {
		cfg.VolumeCount = 1
	}
	ids := cfg.VolumeIDs
	if len(ids) == 0 {
		ids = make([]int32, cfg.VolumeCount)
		for i := 0; i < cfg.VolumeCount; i++ {
			ids[i] = int32(i + 1)
		}
	}
	s := &Store{
		root:    cfg.DataDir,
		vols:    make(map[int32]*volume.Volume),
		maxSize: cfg.VolumeSize,
	}
	for _, id := range ids {
		dir := filepath.Join(cfg.DataDir, fmt.Sprintf("volume_%d", id))
		vol, err := volume.Open(dir, id, volume.Options{MaxSize: cfg.VolumeSize})
		if err != nil {
			_ = s.Close()
			return nil, err
		}
		s.vols[id] = vol
	}
	return s, nil
}

// Write appends data into the given volume.
func (s *Store) Write(vid int32, key int64, cookie int32, data []byte) (meta.NeedleLoc, error) {
	vol, err := s.get(vid)
	if err != nil {
		return meta.NeedleLoc{}, err
	}
	return vol.Write(key, cookie, data)
}

// Read returns needle bytes from volume.
func (s *Store) Read(vid int32, key int64, cookie int32) ([]byte, error) {
	vol, err := s.get(vid)
	if err != nil {
		return nil, err
	}
	return vol.Read(key, cookie)
}

// Delete marks a needle deleted.
func (s *Store) Delete(vid int32, key int64, cookie int32) error {
	vol, err := s.get(vid)
	if err != nil {
		return err
	}
	return vol.Delete(key, cookie)
}

// PickVolume chooses a writable volume with the most free space.
func (s *Store) PickVolume(need int64) (int32, error) {
	s.mu.RLock()
	defer s.mu.RUnlock()
	var best int32
	var bestFree int64 = -1
	for id, vol := range s.vols {
		st := vol.State()
		if st.ReadOnly {
			continue
		}
		if st.FreeSpace < need {
			continue
		}
		if st.FreeSpace > bestFree {
			bestFree = st.FreeSpace
			best = id
		}
	}
	if bestFree < 0 {
		return 0, ErrNoWritable
	}
	return best, nil
}

// States returns all volume states.
func (s *Store) States() []meta.VolumeState {
	s.mu.RLock()
	defer s.mu.RUnlock()
	out := make([]meta.VolumeState, 0, len(s.vols))
	for _, vol := range s.vols {
		out = append(out, vol.State())
	}
	return out
}

func (s *Store) get(vid int32) (*volume.Volume, error) {
	s.mu.RLock()
	defer s.mu.RUnlock()
	vol, ok := s.vols[vid]
	if !ok {
		return nil, ErrVolumeNotFound
	}
	return vol, nil
}

// Close closes all volumes.
func (s *Store) Close() error {
	s.mu.Lock()
	defer s.mu.Unlock()
	var first error
	for id, vol := range s.vols {
		if err := vol.Close(); err != nil && first == nil {
			first = err
		}
		delete(s.vols, id)
	}
	return first
}

File: jfs-store/src/main/java/com/levis9527/jfs/store/Store.java
package com.levis9527.jfs.store;

import com.levis9527.jfs.meta.MetaTypes.NeedleLoc;
import com.levis9527.jfs.meta.MetaTypes.VolumeState;
import com.levis9527.jfs.volume.Volume;
import com.levis9527.jfs.volume.VolumeException;

import java.io.Closeable;
import java.io.IOException;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Physical storage service (bfs store): owns volumes and exposes add/get/del.
 */
public final class Store implements Closeable {
    private final Path root;
    private final Map<Integer, Volume> volumes = new LinkedHashMap<>();

    public Store(Path dataDir, int volumeCount, long volumeSize) throws IOException {
        this.root = dataDir;
        if (volumeCount <= 0) {
            volumeCount = 1;
        }
        for (int i = 1; i <= volumeCount; i++) {
            Path dir = dataDir.resolve("volume_" + i);
            volumes.put(i, new Volume(dir, i, volumeSize));
        }
    }

    public NeedleLoc write(int vid, long key, int cookie, byte[] data) throws IOException {
        return get(vid).write(key, cookie, data);
    }

    public byte[] read(int vid, long key, int cookie) throws IOException {
        return get(vid).read(key, cookie);
    }

    public void delete(int vid, long key, int cookie) throws IOException {
        get(vid).delete(key, cookie);
    }

    public int pickVolume(long needBytes) {
        int best = -1;
        long bestFree = -1;
        for (Volume vol : volumes.values()) {
            VolumeState st = vol.state();
            if (st.readOnly || st.freeSpace < needBytes) {
                continue;
            }
            if (st.freeSpace > bestFree) {
                bestFree = st.freeSpace;
                best = st.id;
            }
        }
        if (best < 0) {
            throw new VolumeException("no writable volume");
        }
        return best;
    }

    public List<VolumeState> states() {
        List<VolumeState> out = new ArrayList<>();
        for (Volume vol : volumes.values()) {
            out.add(vol.state());
        }
        return out;
    }

    private Volume get(int vid) {
        Volume vol = volumes.get(vid);
        if (vol == null) {
            throw new VolumeException("volume not found: " + vid);
        }
        return vol;
    }

    public Path root() {
        return root;
    }

    @Override
    public void close() throws IOException {
        IOException first = null;
        for (Volume vol : volumes.values()) {
            try {
                vol.close();
            } catch (IOException e) {
                if (first == null) {
                    first = e;
                }
            }
        }
        volumes.clear();
        if (first != null) {
            throw first;
        }
    }
}
